daemon/cmd/state.go

```diff
@@ -375,35 +350,59 @@ func (d *Daemon) regenerateRestoredEndpoints(state *endpointRestoreState, endpoi
 	}
 	endpointCleanupCompleted.Wait()
 
+	// Wait on *all* restored endpoints for initial envoy policy computation.
+	// Its safe to do so here because, we expect all endpoints to close this channel,
+	// even if they are eventually deleted and never undergo regeneration.
 	go func() {
 		for _, ep := range state.restored {
 			<-ep.InitialEnvoyPolicyComputed
 		}
 		close(d.endpointInitialPolicyComplete)
 	}()
 
-	go func() {
-		regenerated, total := 0, 0
-		if len(state.restored) > 0 {
-			for buildSuccess := range epRegenerated {
-				if buildSuccess {
-					regenerated++
-				}
-				total++
-				if total >= len(state.restored) {
-					break
-				}
-			}
+	// Trigger regeneration for relevant restored endopints in a separate goroutine.
+	go d.handleRestoredEndpointsRegeneration(endpointsToRegenerate, endpointsRegenerator)
+}
+
+// This method is called during daemon start to asynchronously handle regeneration for restored
+// endpoints. We assume that all the endpoints for which regeneration is requested are
+// already exposed to EndpointManager.
+// This method is responsible for listening/waiting on relevant signals from other subsystems
+// and only performing regeneration for alive endpoints.
```

Comment:
I like the part about the declared assumptions, that context is helpful for readers.

nit: I find the mention of "relevant signals" a bit vague and obscures one of the most important aspects of the function. I'd suggest either deciding this is not relevant in the comment and omitting this sentence, or make it more clear to readers. Suggested alternative below.

One more minor nit to declare the other critical aspect of the operation of this function.

```suggestion
//
// This method assumes that all the endpoints for which regeneration is requested are
// already exposed to EndpointManager. It waits until the Endpoint DeletionQueue is
// drained before regenerating all remaining live endpoints.
//
// Once complete, this method closes the daemon 'endpointRestoreComplete' channel.
```